hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java

```diff
@@ -1726,4 +1659,302 @@ public boolean isInitialized() {
+        LOG.info(String.format(
+            "Cannot compact metadata table as there are %d inflight instants in data table before latest deltacommit in metadata table: %s. Inflight instants in data table: %s",
+            pendingInstants.size(), latestDeltaCommitTimeInMetadataTable, Arrays.toString(pendingInstants.toArray())));
+        return false;
+      }
+
+      // Check if there are any pending compaction or log compaction instants in the timeline.
+      // If pending compact/logCompaction operations are found abort scheduling new compaction/logCompaction operations.
+      Option<HoodieInstant> pendingLogCompactionInstant =
+          metadataMetaClient.getActiveTimeline().filterPendingLogCompactionTimeline().firstInstant();
+      Option<HoodieInstant> pendingCompactionInstant =
+          metadataMetaClient.getActiveTimeline().filterPendingCompactionTimeline().firstInstant();
+      if (pendingLogCompactionInstant.isPresent() || pendingCompactionInstant.isPresent()) {
+        LOG.warn(String.format("Not scheduling compaction or logCompaction, since a pending compaction instant %s or logCompaction %s instant is present",
+            pendingCompactionInstant, pendingLogCompactionInstant));
+        return false;
+      }
+      return true;
+    }
+
+    protected void checkNumDeltaCommits(HoodieTableMetaClient metaClient, int maxNumDeltaCommitsWhenPending) {
+      final HoodieActiveTimeline activeTimeline = metaClient.reloadActiveTimeline();
+      Option<HoodieInstant> lastCompaction = activeTimeline.filterCompletedInstants()
+          .filter(s -> s.getAction().equals(COMMIT_ACTION)).lastInstant();
+      int numDeltaCommits = lastCompaction.isPresent()
+          ? activeTimeline.getDeltaCommitTimeline().findInstantsAfter(lastCompaction.get().requestedTime()).countInstants()
+          : activeTimeline.getDeltaCommitTimeline().countInstants();
+      if (numDeltaCommits > maxNumDeltaCommitsWhenPending) {
+        throw new HoodieMetadataException(String.format("Metadata table's deltacommits exceeded %d: "
+                + "this is likely caused by a pending instant in the data table. Resolve the pending instant "
+                + "or adjust `%s`, then restart the pipeline.",
+            maxNumDeltaCommitsWhenPending, HoodieMetadataConfig.METADATA_MAX_NUM_DELTACOMMITS_WHEN_PENDING.key()));
+      }
+    }
+
+    /**
+     * Perform a compaction on the Metadata Table.
+     * <p>
+     * Cases to be handled:
+     * 1. We cannot perform compaction if there are previous inflight operations on the dataset. This is because
+     * a compacted metadata base file at time Tx should represent all the actions on the dataset till time Tx.
+     * <p>
+     * 2. In multi-writer scenario, a parallel operation with a greater instantTime may have completed creating a
+     * deltacommit.
+     */
+    protected void compactIfNecessary(BaseHoodieWriteClient writeClient, String latestDeltacommitTime) {
```

Comment:
this java docs need to go after L 1775.

Reply:
Java docs for the function `compactIfNecessary` are same in 0.15.0 and 1.0